### PR TITLE
atlasaction: migrate-lint - add comment to PR after linting

### PR DESCRIPTION
**Pull request by dorav**
Creating a PR comment is done by directly calling GitHub API via HTTP.
The github-token input was added since need it to call GitHub API.

### CODE REVIEW

File: atlasaction/action.go [hunk 1]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	ghToken := act.GetInput("github-token")
+	comments, err := g.getIssueComments(prNumber, event.Repository.Name, ghToken)
+	if err != nil {
+		return err
+	}
+	r, err := generateComment(summary, dirName)
+	if err != nil {
+		return err
+	}
+	if ac := findFirst(comments, isAtlasLintCommentFor(dirName)); ac != nil {
```
Comment (a8m) [inline atlasaction/action.go:167]: Use slices.IndexFunc

File: atlasaction/action.go [hunk 2]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
-func createContext(github *githubactions.Action) (*ContextInput, error) {
-	ghContext, err := github.Context()
+type GithubIssueComment struct {
+	Id   int    `json:"id"`
+	Body string `json:"body"`
+}
+
+type GithubAPI struct {
+	baseURL string
+}
```
Comment (a8m) [inline atlasaction/action.go:183]: merge the type definition. Also, If the GithubAPI is public, why the fields are private?

File: atlasaction/action.go [hunk 3]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+		err = g.createIssueComment(prNumber, r, event.Repository.Name, ghToken)
+	}
+
 	return err
 }
 
-func createContext(github *githubactions.Action) (*ContextInput, error) {
-	ghContext, err := github.Context()
+type GithubIssueComment struct {
+	Id   int    `json:"id"`
```
Comment (a8m) [inline atlasaction/action.go:177]:
```suggestion
	ID   int    `json:"id"`
```

File: atlasaction/action.go [hunk 4]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	err = json.Unmarshal(all, &comments)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing github comments with %v/%v from %v, %w", repo, id, string(all), err)
+	}
+	return comments, nil
+}
+
+func (g *GithubAPI) createIssueComment(id int, content io.Reader, repo, authToken string) error {
+	url := fmt.Sprintf("%v/repos/ariga/%v/issues/%v/comments", g.baseURL, repo, id)
+	client := &http.Client{}
```
Comment (a8m) [inline atlasaction/action.go:212]: Use http.DefaultClient

File: atlasaction/action.go [hunk 5]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	if err != nil {
+		return err
+	}
+	addHeaders(req, authToken)
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
```
Comment (a8m) [inline atlasaction/action.go:246]: Use http constants for status codes

File: atlasaction/action.go [hunk 6]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+<strong>Lint report <a href=%q>available here</a></strong>`, dirName, icon, url)
 	act.AddStepSummary(summary)
+
+	g := GithubAPI{
+		baseURL: ghContext.APIURL,
+	}
+	prNumber := event.PullRequestNumber
+	if prNumber == 0 {
+		return fmt.Errorf("unknown pr number")
+	}
```
Comment (a8m) [inline atlasaction/action.go:157]: Is this check relevant? Won't you get an error on report if it does not exist?

File: atlasaction/action.go [hunk 7]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
 	})
 	url := strings.TrimSpace(resp.String())
 	act.SetOutput("report-url", url)
+	publishErr := publishResult(url, err, act)
+	if publishErr != nil {
+		act.Warningf("unable to publish lint report: %v", publishErr)
+	}
```
Comment (a8m) [inline atlasaction/action.go:127]:
```suggestion
	if err1 := publishResult(url, err, act); err1 != nil {
		act.Warningf("unable to publish lint report: %v", err1)
	}
```

File: atlasaction/action.go [hunk 8]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	if err != nil {
+		return err
+	}
+	addHeaders(req, authToken)
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 201 {
```
Comment (a8m) [inline atlasaction/action.go:223]: use http constants

File: atlasaction/action.go [hunk 9]
```diff
@@ -119,37 +121,204 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+		b, err := io.ReadAll(res.Body)
+		if err == nil {
+			err = errors.New(string(b))
+		}
+		return fmt.Errorf("create github comment failed with %v: %w", res.StatusCode, err)
```
Comment (a8m) [inline atlasaction/action.go:228]: Is it really needed besides "unexpected status code: %s"?

File: atlasaction/action.go [hunk 10]
```diff
@@ -119,37 +122,190 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
-	if err := mapstructure.Decode(ghContext.Event, &ev); err != nil {
-		return nil, fmt.Errorf("failed to parse push event: %v", err)
+	defer res.Body.Close()
+	all, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	var comments []githubIssueComment
+	err = json.Unmarshal(all, &comments)
+	if err != nil {
```
Comment (a8m) [inline atlasaction/action.go:203]:
```suggestion
	buf, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var comments []githubIssueComment
	if err = json.Unmarshal(buf, &comments); err != nil {
```

File: atlasaction/action.go [hunk 11]
```diff
@@ -119,37 +122,190 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		b, err := io.ReadAll(res.Body)
+		if err == nil {
+			err = errors.New(string(b))
+		}
```
Comment (a8m) [inline atlasaction/action.go:247]: Why not using the same logic as above?

File: atlasaction/action.go [hunk 12]
```diff
@@ -119,37 +122,190 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+	return err
+}
+
+func addHeaders(req *http.Request, authToken string) {
+	req.Header.Add("Accept", "application/vnd.github+json")
+	req.Header.Add("Authorization", "Bearer "+authToken)
+	req.Header.Add("X-GitHub-Api-Version", "2022-11-28")
+}
+
+func generateComment(data, dir string) (io.Reader, error) {
```
Comment (a8m) [inline atlasaction/action.go:259]: No need for this small functions. Please, inline them.

File: atlasaction/action.go [hunk 13]
```diff
@@ -119,37 +122,190 @@ func MigrateLint(ctx context.Context, client *atlasexec.Client, act *githubactio
...
+
+func triggerEvent(ghContext *githubactions.GitHubContext) (*githubTriggerEvent, error) {
+	var event githubTriggerEvent
+	if err := mapstructure.Decode(ghContext.Event, &event); err != nil {
+		return nil, fmt.Errorf("failed to parse push event: %v", err)
+	}
+	return &event, nil
+}
+
+func isAtlasLintCommentFor(dir string) func(c githubIssueComment) bool {
```
Comment (a8m) [inline atlasaction/action.go:305]: ditto

File: atlasaction/action.go (hunk 11 above)
Comment (dorav) [reply atlasaction/action.go:247]: It should, I missed this, fixing

File: atlasaction/action.go [hunk 14]
```diff
@@ -11,6 +11,9 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"golang.org/x/exp/slices"
```
Comment (a8m) [inline atlasaction/action.go:14]:
```suggestion
	"slices"
```

**Review by a8m — APPROVED** [review]

**Pull request merged**